src/BTrees/BTreeTemplate.c

```diff
@@ -2496,8 +2504,79 @@ static PyNumberMethods BTree_as_number_for_nonzero = {
     bucket_or,                              /* nb_or */
 };
 
-static PyTypeObject BTreeType = {
+static PyObject* BTreeType_setattro_allowed_names; /* initialized in module */
+
+static int
+BTreeType_setattro(PyTypeObject* type, PyObject* name, PyObject* value)
+{
+    /*
+      type.tp_setattro prohibits setting any attributes on a built-in type,
+      so we need to use our own (metaclass) type to handle it. The set of
+      allowable values needs to be carefully controlled.
+
+      Alternately, we could use heap-allocated types when they are supported
+      an all the versions we care about, because those do allow setting attributes.
+    */
+    int allowed;
+    allowed = PySequence_Contains(BTreeType_setattro_allowed_names, name);
+    if (allowed < 0) {
+        return -1;
+    }
+
+    if (allowed) {
+        PyDict_SetItem(type->tp_dict, name, value);
+        PyType_Modified(type);
+        if (PyErr_Occurred()) {
+            return -1;
+        }
+        return 0;
+    }
+    PyErr_Format(
+        PyExc_TypeError,
+        /* distinguish the error message from what type would produce */
+        "BTree: can't set attributes of built-in/extension type '%s'",
+        type->tp_name);
```

Comment:
We can set some attributes. I suggest "cannot set attribute ${name} on type ${type}".

Reply:
Great point! I just [copied what typeobject.c does](https://github.com/python/cpython/blob/e7078ee11667820e3e4585a67d4444d1d469d54e/Objects/typeobject.c#L3401) and stuck "BTree" on the front of it. I'll make a change.